src/cache.rs

```diff
@@ -0,0 +1,495 @@
+use rustc_hash::FxHashMap;
+use serde::Serialize;
+use std::fs;
+use std::fs::File;
+use std::io::{BufReader, BufWriter};
+use std::path::Path;
+use std::result::Result;
+use std::str::FromStr;
+use zksync_basic_types::H256;
+use zksync_types::api::{Block, Transaction, TransactionVariant};
+use zksync_types::Transaction as RawTransaction;
+
+const CACHE_TYPE_BLOCKS_FULL: &str = "blocks_full";
+const CACHE_TYPE_BLOCKS_MIN: &str = "blocks_min";
+const CACHE_TYPE_BLOCK_RAW_TRANSACTIONS: &str = "block_raw_transactions";
+const CACHE_TYPE_TRANSACTIONS: &str = "transactions";
+
+/// Cache configuration. Can be one of:
+///
+/// None    : Caching is disabled
+/// Memory  : Caching is provided in-memory and not persisted across runs
+/// Disk    : Caching is persisted on disk in the provided directory and can be reset
+#[derive(Debug, Clone)]
+pub enum CacheConfig {
+    None,
+    Memory,
+    Disk { dir: String, reset: bool },
+}
+
+impl std::default::Default for CacheConfig {
+    fn default() -> Self {
+        CacheConfig::None
+    }
+}
+
+/// A general purpose cache.
+#[derive(Default, Debug, Clone)]
+pub(crate) struct Cache {
+    config: CacheConfig,
+    block_hashes: FxHashMap<u64, H256>,
+    blocks_full: FxHashMap<H256, Block<TransactionVariant>>,
+    blocks_min: FxHashMap<H256, Block<TransactionVariant>>,
+    block_raw_transactions: FxHashMap<u64, Vec<RawTransaction>>,
+    transactions: FxHashMap<H256, Transaction>,
+}
+
+impl Cache {
+    /// Creates a new cache with the provided config.
+    pub(crate) fn new(config: CacheConfig) -> Self {
+        let mut cache = Cache {
+            config: config.clone(),
+            ..Default::default()
+        };
+
+        if let CacheConfig::Disk { dir, reset } = &config {
+            if *reset {
+                fs::remove_dir_all(Path::new(dir)).unwrap_or_else(|err| {
```

Comment:
I'm hesitant to delete an entire directory given a path in fear that we delete more than is necessary/expected. Can we only remove the directories we plan to create/overwrite (similar to how we create them below)?

```rust
            for cache_type in [
                CACHE_TYPE_BLOCKS_FULL,
                CACHE_TYPE_BLOCKS_MIN,
                CACHE_TYPE_BLOCK_RAW_TRANSACTIONS,
                CACHE_TYPE_TRANSACTIONS,
            ] {
                let cache_path = Path::new(dir).join(cache_type);
                
                if *reset {
                    fs::remove_dir_all(cache_path).unwrap_or_else(|err| {
                                    log::error!("failed removing cache from disk: {:?}", err)
                    });
                }

                fs::create_dir_all(cache_path).unwrap_or_else(|err| {
                    panic!("failed creating directory {}: {:?}", cache_type, err)
                });
            }
```

Reply:
I see your concern and I agree to a certain extent. It's not an impossible scenario that the user passes a wrong directory with `--reset-cache` and in that case I'd prefer that we are less punishing to the user error.